### PR TITLE
Adding MapZipperIterDataPipe

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50
* __->__ #49

Note that there were more discussion that took place [here](https://github.com/pytorch/pytorch/pull/66204).

Need to wait for [this PR](https://github.com/pytorch/pytorch/pull/66275) to land for.

Differential Revision: [D31450886](https://our.internmc.facebook.com/intern/diff/D31450886)

### CODE REVIEW

File: test/_utils/_common_utils_for_test.py [hunk 1]
```diff
@@ -20,6 +21,19 @@ def __iter__(self):
             yield i
 
 
+class MDP(MapDataPipe):
```
[the remark this answers is not in the document]
Comment (NivekT) [reply test/_utils/_common_utils_for_test.py:24]: This is done

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31450886).

File: torchdata/datapipes/iter/util/combining.py [hunk 2]
```diff
@@ -72,3 +68,52 @@ def __iter__(self):
...
+
+    def __init__(
+        self,
+        source_iterdatapipe: IterDataPipe,
+        map_datapipe: MapDataPipe,
+        key_fn: Callable,
+        merge_fn: Callable = tuple_merge,
+    ):
+        self.source_iterdatapipe = source_iterdatapipe
+        self.map_datapipe = map_datapipe
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/combining.py:102]: Could we add a check if it's `MapDataPipe`?

Comment (NivekT) [reply torchdata/datapipes/iter/util/combining.py:104]: Done

**Pull request closed without merging**